### PR TITLE
Fix marginal_counts incorrect creg_sizes with Result object and add format_marginal kwarg

### LINKED ISSUE #5424: Result.utils.marginal_counts returns inconsistent creg structure for multiple cregs
### Information

- **Qiskit Terra version**: master @ d26f273d
- **Python version**: 3.6
- **Operating system**: osx

### Steps to reproduce the problem

```python
from qiskit import *

qr_data = QuantumRegister(3, 'qr_data')
qr_anc = QuantumRegister(2, 'qr_anc')
cr_data = ClassicalRegister(3, 'cr_data')
cr_anc = ClassicalRegister(2, 'cr_anc')

qc = QuantumCircuit(qr_data, qr_anc, cr_data, cr_anc)
qc.x(qr_data)
qc.measure(qr_data, cr_data)
qc.measure(qr_anc, cr_anc)
qc.draw('mpl')
```
image

```python
result = execute(qc, Aer.get_backend('qasm_simulator')).result()
result.get_counts()
```
```
{'00 111': 1024}
```

```python
from qiskit.result.utils import marginal_counts
marginal_counts(result, [0, 1, 2]).get_counts()
```
```
{'11 1': 1024}
```

### PR DESCRIPTION
**Pull request by enavarro51**
### Summary

Fixes #5424

### Details and comments

When a Result object was passed to `result.utils.marginal_counts`, the `memory_slots` value was decremented to match the number of indices passed to `marginal_counts`. The `creg_sizes` though were not decremented, which resulted in incorrect formatting if `get_counts` was called on the result returned from `marginal_counts`. This PR properly decrements the `creg_sizes` to match the indices.

In addition there is a problem with `marginal_counts` in that information as to which indices are used is lost. For example, if there were 2 cregs of size 3 and 2 respectively, this count, `{11 111: 1024}` might be returned from `get_counts`. If `marginal_counts` is called with indices `[0, 2, 4]`, the return value is `{111: 1024}`. This value would be the same for `[0, 1, 2]`, `[1, 3, 4]`, etc. 

An optional `kwarg`, `format_marginal`, has been added to `marginal_counts `which defaults to False. If set to True, underscores are inserted as placeholders in the non-indices positions. For the above example, the output would be `{1_ 1_1: 1024}`. For `[0, 1, 2]`, this would be `{__ 111: 1024}`, and for `[1, 3, 4]` it would be `{11 _1_: 1024}`, allowing differentiation between the different indices inputs. Since these are strings, they will display this way when visualizations like `plot_histogram` are called.

Finally code from @aeddins-ibm in #5241 was used to replace complex and slow code in the `marginalize` function. In basic tests with 20 cregs, the new code was 7 times faster than the old and produced the same output in all the tests.

### CODE REVIEW

**Comment by 1ucian0** [comment]
Adding @aeddins-ibm as a co-author for this PR

File: qiskit/result/utils.py [hunk 1]
```diff
@@ -10,33 +10,35 @@
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
 
-"""Utility functions for working with Results."""
+"""Utility functions for working with Result counts."""
```
Comment (1ucian0) [inline qiskit/result/utils.py:13]: I guess this was referring to `Result` as the class. Given that `Counts` is a related-but-different class, I think this change is more confusing that helping.

File: qiskit/result/utils.py [hunk 2]
```diff
@@ -52,63 +54,84 @@ def marginal_counts(result, indices=None, inplace=False):
...
-    # Check if we do not need to marginalize. In this case we just trim
-    # whitespace from count keys
+    # Check if we do not need to marginalize and if so, trim
+    # whitespace and '_' and return
     if (indices is None) or set(range(num_clbits)) == set(indices):
         ret = {}
         for key, val in counts.items():
-            key = key.replace(' ', '')
+            key = key.replace(' ', '').replace('_', '')
             ret[key] = val
```
Comment (1ucian0) [inline qiskit/result/utils.py:103]: This pattern repeats a lot. Maybe worthy to make a function (in `counts.py`, I assume) for it?

**Review by 1ucian0 — CHANGES_REQUESTED** [review]
looks good! just some comments

File: qiskit/result/utils.py (hunk 1 above)
Comment (enavarro51) [reply qiskit/result/utils.py]: Makes sense. Done.

File: qiskit/result/utils.py (hunk 2 above)
Comment (enavarro51) [reply qiskit/result/utils.py]: Added a function to both counts.py (3 uses) and utils.py (2 uses).

**Review by 1ucian0 — APPROVED** [review]
Thanks!

**Pull request merged**